src/uct/rocm/copy/rocm_copy_md.c

```diff
@@ -46,48 +53,106 @@ static ucs_status_t uct_rocm_copy_md_query(uct_md_h md, uct_md_attr_t *md_attr)
 {
-    *rkey_p   = 0xdeadbeef;
+    uct_rocm_copy_key_t *packed = (uct_rocm_copy_key_t *)rkey_buffer;
+    uct_rocm_copy_key_t *key;
+
+    key = ucs_malloc(sizeof(uct_rocm_copy_key_t), "uct_rocm_copy_key_t");
+    if (NULL == key) {
+        ucs_error("failed to allocate memory for uct_rocm_copy_key_t");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    key->vaddr      = packed->vaddr;
+    key->dev_ptr    = packed->dev_ptr;
+
     *handle_p = NULL;
+    *rkey_p   = (uintptr_t)key;
+
     return UCS_OK;
 }
 
 static ucs_status_t uct_rocm_copy_rkey_release(uct_component_t *component,
                                                uct_rkey_t rkey, void *handle)
 {
+    ucs_assert(NULL == handle);
+    ucs_free((void *)rkey);
     return UCS_OK;
 }
 
-static ucs_status_t uct_rocm_copy_mem_reg(uct_md_h md, void *address, size_t length,
-                                          unsigned flags, uct_mem_h *memh_p)
+static ucs_status_t uct_rocm_copy_mem_reg_internal(
+        uct_md_h uct_md, void *address, size_t length,
+        unsigned flags, uct_rocm_copy_mem_t *mem_hndl)
 {
+
     hsa_status_t status;
-    void *lock_addr;
+    void *dev_addr = address;
 
     if(address == NULL) {
-        *memh_p = address;
+        memset(mem_hndl, 0, sizeof(*mem_hndl));
         return UCS_OK;
     }
 
-    status = hsa_amd_memory_lock(address, length, NULL, 0, &lock_addr);
+    status = hsa_amd_memory_lock(address, length, NULL, 0, &dev_addr);
     if (status != HSA_STATUS_SUCCESS) {
         return UCS_ERR_IO_ERROR;
     }
 
-    *memh_p = address;
+    mem_hndl->vaddr    = address;
+    mem_hndl->dev_ptr  = dev_addr;
+    mem_hndl->reg_size = length;
+
+    ucs_trace("Registered addr %p len %zu dev addr %p", address, length, dev_addr);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_rocm_copy_mem_reg(uct_md_h md, void *address, size_t length,
+                                          unsigned flags, uct_mem_h *memh_p)
+{
+    uct_rocm_copy_mem_t *mem_hndl = NULL;
+    void *start, *end;
+    size_t len, page_size;
+    ucs_status_t status;
+
+    mem_hndl = ucs_malloc(sizeof(uct_rocm_copy_mem_t), "rocm_copy handle");
+    if (NULL == mem_hndl) {
+        ucs_error("failed to allocate memory for rocm_copy_mem_t");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    page_size = ucs_get_page_size();
+    start = ucs_align_down_pow2_ptr(address, page_size);
+    end   = ucs_align_up_pow2_ptr(UCS_PTR_BYTE_OFFSET(address, length), page_size);
+    len   = UCS_PTR_BYTE_DIFF(start, end);
+    ucs_assert_always(start <= end);
+
+    status = uct_rocm_copy_mem_reg_internal(md, address, len, 0, mem_hndl);
+    if (status != UCS_OK) {
+        ucs_free(mem_hndl);
+        return status;
+    }
+
+    *memh_p = mem_hndl;
     return UCS_OK;
 }
 
 static ucs_status_t uct_rocm_copy_mem_dereg(uct_md_h md, uct_mem_h memh)
 {
-    void *address = (void *)memh;
     hsa_status_t status;
+    uct_rocm_copy_mem_t *mem_hndl    = (uct_rocm_copy_mem_t *)memh;
```

Comment:
spaces before =

Reply:
Sorry about missing these. I have updated the PR.